src/github.com/matrix-org/dendrite/clientapi/routing/register.go

```diff
@@ -387,23 +209,19 @@ func Register(
 		sessionID = util.RandomString(sessionIDLength)
 	}
 
-	// If no auth type is specified by the client, send back the list of available flows
-	if r.Auth.Type == "" {
-		return util.JSONResponse{
-			Code: http.StatusUnauthorized,
-			JSON: newUserInteractiveResponse(sessionID,
-				cfg.Derived.Registration.Flows, cfg.Derived.Registration.Params),
-		}
+	if cfg.Matrix.RegistrationDisabled && r.Auth.Type != authtypes.LoginTypeSharedSecret {
+		return util.MessageResponse(http.StatusForbidden, "Registration has been disabled")
 	}
 
 	// Squash username to all lowercase letters
 	r.Username = strings.ToLower(r.Username)
 
-	if resErr = validateUserName(r.Username); resErr != nil {
+	if resErr = validateCredentials(r.Username, r.Password); resErr != nil {
 		return *resErr
 	}
-	if resErr = validatePassword(r.Password); resErr != nil {
-		return *resErr
+
+	if res := handleLoginTypes(req, r, cfg, sessionID); res != nil {
```

Comment:
again, why is this happening here, in the *middle* of the validation of the username and password parameters?

I suggest you leave it in `handleRegistrationFlow`.